DOC update and improve the `sample_weight` entry in the glossary (#30564)
scikit-learn/scikit-learn: doc/glossary.rst

        their selection for training (e.g. inverse propensity weighting). It is
        also useful to model the frequency of an event of interest per unit of
        time on a dataset of observations with different exposure durations per
        individual (see
        :ref:`sphx_glr_auto_examples_linear_model_plot_poisson_regression_non_normal_loss.py`
        and
        :ref:`sphx_glr_auto_examples_linear_model_plot_tweedie_regression_insurance_claims.py`).

        Some model hyper-parameters are expressed in terms of a discrete number
        of samples in a region of the feature space. When fitting with sample
        weights, a count of samples is often automatically converted to a sum
        of their weights, but this is not always the case. Please refer to the
        docstring for details.

        In classification, weights can also be specified for all samples
        belonging to a given target class with the :term:`class_weight`
        estimator :term:`parameter`. If both ``sample_weight`` and
        ``class_weight`` are provided, the final weight assigned to a sample is
        the product of the two.

>         At the time of writing (version 1.7), not all scikit-learn estimators

Author: Since 1.7 was released, this doc will be published on scikit-learn.org/stable no sooner than 1.8.

```suggestion
        At the time of writing (version 1.8), not all scikit-learn estimators
```

        See :ref:`group_cv`.

    ``sample_weight``
        A relative weight for each sample. Intuitively, if all weights are
        integers, using them in a model or scorer is like duplicating each
        sample as many times as the weight value. Weights may be specified as
        floats, so that sample weights are usually equivalent up to a constant

>         positive scaling factor.

Reviewer: equivalent to what? I don't understand this sentence.

Author: Invariance to the scale of weights is actually not guaranteed for all estimators or hyperparameter choices. So let's be more generic.

```suggestion
        floats to express fractional relative importance of data points with respect
        to one another others.
```

Author: I did not want to explicitly state that because we are not yet sure to which extent this is true or not: we do haven't started testing this aspect systematically, but we know that this is not the case for many important ones such as `LogisticRegression` and `Ridge`. Both those estimators have a regularization parameter whose effect depends on the scale of the weights.

        <https://en.wikipedia.org/wiki/Inverse_probability_weighting>`_ of
        their selection for training (e.g. inverse propensity weighting). It is
        also useful to model the frequency of an event of interest per unit of
        time on a dataset of observations with different exposure durations per
        individual (see
        :ref:`sphx_glr_auto_examples_linear_model_plot_poisson_regression_non_normal_loss.py`
        and
        :ref:`sphx_glr_auto_examples_linear_model_plot_tweedie_regression_insurance_claims.py`).

        Third-party libraries can also use `sample_weight`-compatible
        estimators as building blocks to reduce a specific statistical task
        into a weighted regression or classification task. For instance sample
        weights can be constructed to adjust a time-to-event model for
        censoring in a predictive survival analysis setting. In causal
        inference, it is possible to reduce a conditional average treatment
        effect estimation task to a weighted regression task under some
        assumptions. Sample weights can also be used to mitigate
        fairness-related harms based on a given quantitative definition of

>         fairness.

Reviewer: This paragraph doesn't feel like it belongs here.

Author: Personally, I find this paragraph very important: it grounds an abstract concept into practical application cases. Have those use cases in mind is helpful both for users and contributors.

        Some model hyper-parameters are expressed in terms of a discrete number
        of samples in a region of the feature space. When fitting with sample
        weights, a count of samples is often automatically converted to a sum
        of their weights as is the case  for `min_samples` in
        :class:`cluster.DBSCAN`, for instance. However, this is not always the
        case. In particular, the ``min_samples_leaf`` parameter in
        :class:`ensemble.RandomForestClassifier` does not take weights into
        account. One should instead pass `min_weight_fraction_leaf` to
        :class:`ensemble.RandomForestClassifier` to specify the minimum sum of

>         weights of samples in a leaf.

Reviewer: I think this information is best placed in the docstrings of those classes, not here, or here keep only the first 3-4 lines

Author: ```suggestion
        Some model hyper-parameters are expressed in terms of a discrete number
        of samples in a region of the feature space. When fitting with sample
        weights, a count of samples is often automatically converted to a sum
        of their weights, but this is not always the case. Please refer to the
        docstring for details.
```

        the weight.  Weights may be specified as floats, so that sample weights
        are usually equivalent up to a constant positive scaling factor.

        Weighting samples can be useful in several contexts. For instance, if the
        training data is not uniformly sampled from the target population, it
        is possible to correct for this by weighting training data points

>         according to the inverse probability of being sampled for training. It

Reviewer 2: "[...], it can be corrected by weighting the training data points based on the inverse probability of their selection for training."

        is also useful to model the frequency of an event of interest per unit
        of time on a dataset of observations with different exposure durations
        per individual (see
        :ref:`sphx_glr_auto_examples_linear_model_plot_poisson_regression_non_normal_loss.py`
        and
        :ref:`sphx_glr_auto_examples_linear_model_plot_tweedie_regression_insurance_claims.py`).

>         Third-party libraries can also use `sample_weight`-capable estimators

Reviewer 2: "Third-party libraries can also use `sample_weight`-compatible estimators"

        :class:`ensemble.RandomForestClassifier` does not take weights into
        account. One should instead pass `min_weight_fraction_leaf` to
        :class:`ensemble.RandomForestClassifier` to specify the minimum sum of
        weights of samples in a leaf.

        In classification, weights can also be specified for all samples
        belonging to a given target class with the :term:`class_weight`
        estimator :term:`parameter`. If both ``sample_weight`` and
        ``class_weight`` are provided, the final weight assigned to a sample is
        the product of the two.

        `sample_weight` can be both an argument of the estimator's `fit` method
        for model training or a parameter of a :term:`scorer` for model
        evaluation.

        At the time of writing, not all scikit-learn estimators correctly
        implement the  weight-repetition equivalence property. The `#16298 meta
        issue <https://github.com/scikit-learn/scikit-learn/issues/16298>`_
        tracks ongoing work to detect and fix remaining discrepancies.

        Furthermore, some estimators have a stochastic fit method. For

>         instance, k-means clustering depends on a random initialization,

Reviewer 2: :class:`cluster.KMeans`

>         the weight.  Weights may be specified as floats, so that sample weights

Reviewer 3: The second sentence ready rather complicated and can be simplified.

Maybe like this:

`Intuitively, if all weights are integers, using them in a model or scorer is like duplicating each sample as many times as the weight value.`

        Some model hyper-parameters are expressed in terms of a discrete number
        of samples in a region of the feature space. When fitting with sample
        weights, a count of samples is often automatically converted to a sum
        of their weights as is the case  for `min_samples` in
        :class:`cluster.DBSCAN`, for instance. However, this is not always the
        case. In particular, the ``min_samples_leaf`` parameter in
        :class:`ensemble.RandomForestClassifier` does not take weights into
        account. One should instead pass `min_weight_fraction_leaf` to
        :class:`ensemble.RandomForestClassifier` to specify the minimum sum of
        weights of samples in a leaf.

        In classification, weights can also be specified for all samples
        belonging to a given target class with the :term:`class_weight`
        estimator :term:`parameter`. If both ``sample_weight`` and
        ``class_weight`` are provided, the final weight assigned to a sample is
        the product of the two.

        `sample_weight` can be both an argument of the estimator's `fit` method
        for model training or a parameter of a :term:`scorer` for model

>         evaluation.

Reviewer 3: I think this should be one of the first few sentences in this glossary entry.

In accordance with the rest of our terminology, we should talk about these methods _consuming_ sample_weight and mention that there are others that are able to _route_ sample_weight (with a link-reference to metadata routing).

>         At the time of writing, not all scikit-learn estimators correctly

Reviewer 3: ```suggestion
        At the time of writing (version 1.7), not all scikit-learn estimators correctly
```

>         implement the  weight-repetition equivalence property. The `#16298 meta

Author: ```suggestion
        implement the weight-repetition equivalence property. The `#16298 meta
```

        issue <https://github.com/scikit-learn/scikit-learn/issues/16298>`_
        tracks ongoing work to detect and fix remaining discrepancies.

        Furthermore, some estimators have a stochastic fit method. For
        instance, :class:`cluster.KMeans` depends on a random initialization,
        bagging models randomly resample from the training data, etc. In this
        case, the sample weight-repetition equivalence property described above
        does not hold exactly. However, it should hold at least in expectation

>         over the randomness of the fitting procedure.

Author: Note to reviewers: I think the equivalence should even hold in distribution rather than just in expectation, but I am not 100% sure this can always be enforced in scikit-learn (yet).

        See :ref:`group_cv`.

    ``sample_weight``
        A weight for each data point. Intuitively, if all weights are integers,
        using them in an estimator or a :term:`scorer` is like duplicating each
        data point as many times as the weight value. Weights may also be
        specified as floats: many estimators and scorers are invariant to a
        rescaling of the weights by a constant positive factor but there are

>         exceptions.

Reviewer 4: Sorry but I don't quite follow how "Weights may also be specified as floats" relates to "many estimators and scorers are invariant to a rescaling of the weights by a constant positive factor but there are exceptions".

Author: Because floating-point valued weights could be used to express relative weights as opposed to absolute (integer) repetitions if you divide integer weights by their sum across the training set, for instance. However, there are known case where this is not equivalent (and we are still not sure if it's a bug or a feature).

Reviewer 4: I *think* I see what you are saying.
Integer weights could also be expressed by a float e.g., float 'relative' weights of [0.1, 0.2, 0.3], can be rescaled by a constant factor of 10 and would become integer weights [1, 2, 3] ?

I guess if weights were scale invariant, the above weights would effectively be the same.

I was recently reading this conversation: https://github.com/numpy/numpy/issues/8935#issuecomment-1398689411 (it's about weights specifically in quantiles) but it made me question if some interpretations of weight *is* sample size dependent (i.e. scale variant)..?

Sorry, this is probably not helpful in moving this forward.

Author: Yes, the point is that many scikit-learn estimators have a `fit` method that is naturally sample_weight-scale invariant, but there are exceptions (sometimes only for some combinations of constructor param values). We would need to conduct an exhaustive survey similarly to what we did to check the validity of the integer frequency semantics.

For estimators that are weight-scale invariant, the semantics of fractional weights (passed as floating point values) is easy to define: they are the same as the integer frequency weights because dividing or multiplying by an arbitrary constant that does not impact the results.

For estimators that are not weight scale invariant, the semantics of non-integer weights is not easy to define, but I don't want to make the glossary entry too complex.

>         `sample_weight` can be both an argument of the estimator's `fit` method

Reviewer 4: nitpick

```suggestion
        `sample_weight` can be both an argument of the estimator's :term:`fit` method
```

        `inverse probability
        <https://en.wikipedia.org/wiki/Inverse_probability_weighting>`_ of
        their selection for training (e.g. inverse propensity weighting).

        Some model hyper-parameters are expressed in terms of a discrete number
        of data points in a region of the feature space. When fitting with
        sample weights, a count of data points is often automatically converted
        to a sum of their weights, but this is not always the case. Please
        refer to the model docstring for details.

        In classification, weights can also be specified for all samples
        belonging to a given target class with the :term:`class_weight`
        estimator :term:`parameter`. If both ``sample_weight`` and
        ``class_weight`` are provided, the final weight assigned to a sample is
        the product of the two.

        At the time of writing (version 1.8), not all scikit-learn estimators
        correctly implement the weight-repetition equivalence property. The
        `#16298 meta issue
        <https://github.com/scikit-learn/scikit-learn/issues/16298>`_ tracks

>         ongoing work to detect and fix remaining discrepancies.

Reviewer 4: The 'see below' was referring to this part, but I think this is fine as is!
